File: projectforge-wicket/src/test/java/org/projectforge/core/CreateI18nKeys.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.core;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.projectforge.business.address.FormOfAddress;
import org.projectforge.business.fibu.AuftragsStatus;
import org.projectforge.common.i18n.Priority;
import org.projectforge.framework.calendar.MonthHolder;
import org.projectforge.framework.time.DayHolder;
import org.projectforge.web.MenuItemDef;
import org.projectforge.web.MenuItemDefId;
import org.projectforge.web.wicket.WebConstants;

/**
 * Tries to get all used i18n keys from the sources (java and html). As result a file is written which will be checked
 * by AdminAction.checkI18nProperties. Unused i18n keys should be detected.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
public class CreateI18nKeys
{
  // TODO: I18nEnums

  private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(CreateI18nKeys.class);

  private static final String PATH = "src/main/";

  private static final String PATH_DAYHOLDER = getPathForClass(DayHolder.class);

  private static final String PATH_MONTHHOLDER = getPathForClass(MonthHolder.class);

  private static final String PATH_FORM_OF_ADDRESS = getPathForClass(FormOfAddress.class);

  private static final String PATH_MENU_ITEM_DEF = getPathForClass(MenuItemDef.class);

  private static final String PATH_AUFTRAG_STATUS = getPathForClass(AuftragsStatus.class);

  private static final String PATH_PRIORITY = getPathForClass(Priority.class);

  private static final String I18N_KEYS_FILE = "src/main/resources/" + WebConstants.FILE_I18N_KEYS;

  public static void main(final String[] args) throws IOException
  {
    new CreateI18nKeys().run();
  }

  public void run() throws IOException
  {
    log.info("Create file with all detected i18n keys.");
    final Map<String, Set<String>> i18nKeyUsage = new HashMap<String, Set<String>>();
    parseHtml(i18nKeyUsage);
    parseJava(i18nKeyUsage);
    final FileWriter writer = new FileWriter(I18N_KEYS_FILE);
    writer
        .append(
            "# Don't edit this file. This file is only for developers for checking i18n keys and detecting missed and unused ones.\n");
    final Set<String> i18nKeys = new TreeSet<String>(i18nKeyUsage.keySet());
    for (final String i18nKey : i18nKeys) {
      writer.append(i18nKey).append("=");
      final Set<String> set = i18nKeyUsage.get(i18nKey);
      boolean first = true;
      for (final String filename : set) {
        if (first == false) {
          writer.append(',');
        } else {
          first = false;
        }
        writer.append(filename);
      }
      writer.append("\n");
    }
    IOUtils.closeQuietly(writer);
    log.info("Creation of file of found i18n keys done: " + I18N_KEYS_FILE);
  }

  private void parseHtml(final Map<String, Set<String>> i18nKeyUsage) throws IOException
  {
    final Collection<File> files = listFiles(PATH + "java", "html");
    for (final File file : files) {
      final String content = getContent(file);
      find(file, i18nKeyUsage, content, "<wicket:message\\s+key=\"([a-zA-Z0-9\\.]+)\"\\s/>");
    }
  }

  private void parseJava(final Map<String, Set<String>> i18nKeyUsage) throws IOException
  {
    final Collection<File> files = listFiles(PATH, "java");
    for (final File file : files) {
      final String content = getContent(file);
      find(file, i18nKeyUsage, content, "getString\\(\"([a-zA-Z0-9\\.]+)\"\\)"); // getString("i18nKey")
      find(file, i18nKeyUsage, content, "getLocalizedString\\(\"([a-zA-Z0-9\\.]+)\"\\)"); // getLocalizedString("i18nkey")
      find(file, i18nKeyUsage, content, "getLocalizedMessage\\(\"([a-zA-Z0-9\\.]+)\""); // getLocalizedMessage("i18nkey"
      find(file, i18nKeyUsage, content, "addError\\(\"([a-zA-Z0-9\\.]+)\"\\)"); // addError("i18nkey")
      find(file, i18nKeyUsage, content, "addError\\([a-zA-Z0-9_\"\\.]+,\\s+\"([a-zA-Z0-9\\.]+)\""); // addError(..., "i18nkey"
      find(file, i18nKeyUsage, content, "addGlobalError\\(\"([a-zA-Z0-9\\.]+)\"\\)"); // addGlobalError("i18nkey")
      find(file, i18nKeyUsage, content, "resolveMessage\\(\"([a-zA-Z0-9\\.]+)\"\\)"); // resolveMessage("i18nkey")
      find(file, i18nKeyUsage, content, "throw new UserException\\(\"([a-zA-Z0-9\\.]+)\""); // throw new UserException("i18nkey"
      find(file, i18nKeyUsage, content, "throw new AccessException\\(\"([a-zA-Z0-9\\.]+)\""); // throw new AccessException("i18nkey"
      find(file, i18nKeyUsage, content, "I18N_KEY_[A-Z0-9_]+ = \"([a-zA-Z0-9\\.]+)\""); // I18N_KEY_... = "i18nKey"
      find(file, i18nKeyUsage, content, "new Holiday\\(\"([a-zA-Z0-9\\.]+)\""); // new Holiday("i18nKey"
      find(file, i18nKeyUsage, content,
          "MessageAction.getForwardResolution\\([a-zA-Z0-9_\"\\.]+,\\s+\"([a-zA-Z0-9\\.]+)\"\\)"); // MessageAction.getForwardResolution(...,
      // "i18nkey");
      if (file.getPath().endsWith(PATH_DAYHOLDER) == true) {
        for (final String key : DayHolder.DAY_KEYS) {
          add(i18nKeyUsage, "calendar.day." + key, file);
          add(i18nKeyUsage, "calendar.shortday." + key, file);
        }
      } else if (file.getPath().endsWith(PATH_MONTHHOLDER) == true) {
        for (final String key : MonthHolder.MONTH_KEYS) {
          add(i18nKeyUsage, "calendar.month." + key, file);
        }
      } else if (file.getPath().endsWith(PATH_FORM_OF_ADDRESS) == true) {
        for (final FormOfAddress form : FormOfAddress.values()) {
          add(i18nKeyUsage, form.getI18nKey(), file);
        }
      } else if (file.getPath().endsWith(PATH_AUFTRAG_STATUS) == true) {
        for (final AuftragsStatus status : AuftragsStatus.values()) {
          add(i18nKeyUsage, status.getI18nKey(), file);
        }
      } else if (file.getPath().endsWith(PATH_PRIORITY) == true) {
        for (final Priority priority : Priority.values()) {
          add(i18nKeyUsage, "priority." + priority.getKey(), file);
        }
      } else if (file.getPath().endsWith(PATH_MENU_ITEM_DEF) == true) {
        for (final MenuItemDefId menuItem : MenuItemDefId.values()) {
          add(i18nKeyUsage, menuItem.getI18nKey(), file);
        }
      } else if (file.getName().endsWith("Page.java") == true
          && (content.contains("extends AbstractListPage") == true
          || content.contains("extends AbstractEditPage") == true)) { // Wicket
        // Page
        List<String> list = find(file, content, "super\\(parameters, \"([a-zA-Z0-9\\.]+)\"\\);"); // super(parameters, "i18nKey");
        for (final String entry : list) {
          add(i18nKeyUsage, entry + ".title.edit", file);
          add(i18nKeyUsage, entry + ".title.list", file);
          add(i18nKeyUsage, entry + ".title.list.select", file);
        }
        list = find(file, content, "super\\(caller, selectProperty, \"([a-zA-Z0-9\\.]+)\"\\);"); // super(caller, selectProperty,
        // "i18nKey");
        for (final String entry : list) {
          add(i18nKeyUsage, entry + ".title.edit", file);
          add(i18nKeyUsage, entry + ".title.list", file);
          add(i18nKeyUsage, entry + ".title.list.select", file);
        }
      }
    }
  }

  private static String getPathForClass(final Class<?> clazz)
  {
    return clazz.getName().replace(".", "/") + ".java";
  }

  private void find(final File file, final Map<String, Set<String>> i18nKeyUsage, final String content,
      final String regexp)
  {
    find(file, i18nKeyUsage, content, regexp, null);
  }

  private void find(final File file, final Map<String, Set<String>> i18nKeyUsage, final String content,
      final String regexp,
      final String prefix)
  {
    final List<String> list = find(file, content, regexp);
    for (final String entry : list) {
      final String key = prefix != null ? prefix + entry : entry;
      add(i18nKeyUsage, key, file);
    }
  }

  private List<String> find(final File file, final String content, final String regexp)
  {
    final List<String> result = new ArrayList<String>();
    final Pattern p = Pattern.compile(regexp, Pattern.MULTILINE); // Compiles regular expression into Pattern.
    final Matcher m = p.matcher(content);
    while (m.find()) {
      result.add(m.group(1));
    }
    return result;
  }

  private void add(final Map<String, Set<String>> i18nKeyUsage, final String key, final File file)
  {
    Set<String> set = i18nKeyUsage.get(key);
    if (set == null) {
      set = new TreeSet<String>();
      i18nKeyUsage.put(key, set);
    }
    set.add(file.getPath());
  }

  private Collection<File> listFiles(final String path, final String suffix)
  {
    return FileUtils.listFiles(new File(path), new String[] { suffix }, true);
  }

  private String getContent(final File file) throws IOException
  {
    return FileUtils.readFileToString(file, "UTF-8");
  }
}

File: projectforge-application/src/test/kotlin/org/projectforge/CreateI18nKeys.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2022 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge

import de.micromata.merlin.excel.ExcelWorkbook
import mu.KotlinLogging
import org.apache.commons.io.FileUtils
import org.projectforge.business.book.BookStatus
import org.projectforge.business.user.UserRightId
import org.projectforge.common.anots.PropertyInfo
import org.projectforge.common.i18n.I18nEnum
import org.projectforge.framework.calendar.MonthHolder
import org.projectforge.framework.persistence.api.IUserRightId
import org.projectforge.framework.time.DayHolder
import org.projectforge.menu.builder.MenuItemDef
import org.projectforge.menu.builder.MenuItemDefId
import org.projectforge.web.wicket.WebConstants
import org.reflections.Reflections
import org.reflections.scanners.Scanners
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.regex.Pattern
import kotlin.io.path.name
import kotlin.streams.toList


fun main() {
  CreateI18nKeys().run()
}

private val log = KotlinLogging.logger {}

/**
 * Tries to get all used i18n keys from the sources (java and html). As result a file is written which will be checked
 * by AdminAction.checkI18nProperties. Unused i18n keys should be detected.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
class CreateI18nKeys {
  private class I18nKeyInfo(val i18nKey: String) {
    var translation: String? = null
    var translationDE: String? = null
    val locationList = mutableListOf<String>()
    fun addLocation(location: String) {
      locationList.add(location)
    }

    val locations: String
      get() = locationList.sorted().joinToString()
  }

  private val i18nKeyMap = mutableMapOf<String, I18nKeyInfo>()

  private val srcMainDirs: List<Path>
  private val resourceBundleNames = mutableListOf<String>()
  private val resourceBundles = mutableListOf<ResourceBundle>()
  private val resourceBundlesDE = mutableListOf<ResourceBundle>()
  private val reflections = Reflections("org.projectforge", Scanners.values())

  private var htmlTemplatesCounter = 0

  init {
    srcMainDirs = Files.walk(Paths.get(basePath), 4) // plugins has depth 4
      .filter { Files.isDirectory(it) && it.name == "main" && it.parent?.name == "src" }.toList()
    srcMainDirs.forEach { main ->
      val resourcesDir = Files.walk(main, 1) // src/main/resourcres/***I18n*.properties
        .filter { Files.isDirectory(it) && it.name == "resources" }.toList()
      if (resourcesDir.isNotEmpty()) {
        val propertiesFiles = listFiles(resourcesDir[0], "properties")
        propertiesFiles.filter { it.name.endsWith("_en.properties") }.forEach {
          resourceBundleNames.add(it.name.removeSuffix("_en.properties"))
        }
      }
    }
    resourceBundleNames.sort()
    resourceBundleNames.forEach {
      resourceBundles.add(ResourceBundle.getBundle(it, Locale.ENGLISH))
      resourceBundlesDE.add(ResourceBundle.getBundle(it, Locale.GERMAN))
    }
    resourceBundles.forEach { bundle ->
      bundle.keySet().forEach { key ->
        ensureI18nKeyInfo(key).translation = bundle.getString(key)
      }
    }
    resourceBundlesDE.forEach { bundle ->
      bundle.keySet().forEach { key ->
        ensureI18nKeyInfo(key).translationDE = bundle.getString(key)
      }
    }
    println(resourceBundleNames)
  }

  @Throws(IOException::class)
  fun run() {
    log.info("Create file with all detected i18n keys.")
    srcMainDirs.forEach { path ->
      parseJava(path)
      parseKotlin(path)
      parseWicketHtml(path)
      parseHtmlMailTemplates(path)
    }
    getI18nEnums()
    getPropertyInfos()
    val workbook = ExcelWorkbook.createEmptyWorkbook()
    val sheet = workbook.createOrGetSheet("I18n keys")
    sheet.registerColumn("I18n key|60", "i18nKey")
    sheet.registerColumn("English|60", "translation")
    sheet.registerColumn("German|60", "translationDE")
    sheet.registerColumn("Locations|60")
    sheet.createRow().fillHeadRow()
    i18nKeyMap.keys.sorted().forEach { key ->
      val row = sheet.createRow()
      row.autoFillFromObject(i18nKeyMap[key])
    }
    sheet.setAutoFilter()
    val file = File("i18nKeys.xlsx")
    workbook.asByteArrayOutputStream.use { baos ->
      file.writeBytes(baos.toByteArray())
    }
    println("File '${file.absolutePath}' written.")
    println("Matching html mail templates: $htmlTemplatesCounter")
  }

  @Throws(IOException::class)
  private fun parseWicketHtml(path: Path) {
    val files = listFiles(path, "html")
    for (file in files) {
      val content = getContent(file)
      find(file, content, "<wicket:message\\s+key=\"([a-zA-Z0-9\\.]+)\"\\s/>")
    }
  }

  @Throws(IOException::class)
  private fun parseHtmlMailTemplates(path: Path) {
    val files = listFiles(path, "html")
    for (file in files) {
      val content = getContent(file)
      find(content, "pf.getI18nString\\(\"([a-zA-Z0-9\\.]+)\"\\)").forEach {
        ++htmlTemplatesCounter
        add(it, file)
      }
    }
  }

  private fun getPropertyInfos() {
    // val annotations = reflections.getTypesAnnotatedWith(PropertyInfo::class.java)
    reflections.getFieldsAnnotatedWith(PropertyInfo::class.java)?.forEach { field ->
      val ann = field.getAnnotation(PropertyInfo::class.java)
      addAnnotationKeys(ann.i18nKey, ann.additionalI18nKey, field.declaringClass)
    }
    reflections.getMethodsAnnotatedWith(PropertyInfo::class.java)?.forEach { method ->
      val ann = method.getAnnotation(PropertyInfo::class.java)
      addAnnotationKeys(ann.i18nKey, ann.additionalI18nKey, method.declaringClass)
    }
  }

  private fun addAnnotationKeys(i18nKey: String?, additionalI18nKey: String?, declaringClass: Class<*>) {
    if (!i18nKey.isNullOrBlank()) {
      add(i18nKey, declaringClass.simpleName)
    }
    if (!additionalI18nKey.isNullOrBlank()) {
      add(additionalI18nKey, declaringClass.simpleName)
    }
  }

  private fun getI18nEnums() {
    val subTypes = reflections.getSubTypesOf(I18nEnum::class.java)
    subTypes.forEach {
      if (IUserRightId::class.java.isAssignableFrom(it)) {
        // OK, ignore IUserRightId
      } else {
        it.enumConstants.let { enumConstants ->
          if (enumConstants != null) {
            BookStatus.DISPOSED.i18nKey
            if (it.isEnum) {
              it.enumConstants.forEach { enum ->
                add("${enum.i18nKey}", it.simpleName)
              }
            }
          } else {
            println("******************* Oups, enumConstants are null: ${it.name}")
          }
        }
      }
    }
    UserRightId.values().forEach {
      add("${it.i18nKey}", UserRightId::class.java.simpleName)
    }
  }

  @Throws(IOException::class)
  private fun parseJava(path: Path) {
    val files = listFiles(path, "java")
    for (file in files) {
      val content = getContent(file)
      find(file, content, "getString\\(\"([a-zA-Z0-9\\.]+)\"\\)") // getString("i18nKey")
      find(
        file,
        content,
        "getLocalizedString\\(\"([a-zA-Z0-9\\.]+)\"\\)"
      ) // getLocalizedString("i18nkey")
      find(
        file,
        content,
        "getLocalizedMessage\\(\"([a-zA-Z0-9\\.]+)\""
      ) // getLocalizedMessage("i18nkey"
      find(file, content, "addError\\(\"([a-zA-Z0-9\\.]+)\"\\)") // addError("i18nkey")
      find(
        file,
        content,
        "addError\\([a-zA-Z0-9_\"\\.]+,\\s+\"([a-zA-Z0-9\\.]+)\""
      ) // addError(..., "i18nkey"
      find(file, content, "addGlobalError\\(\"([a-zA-Z0-9\\.]+)\"\\)") // addGlobalError("i18nkey")
      find(file, content, "resolveMessage\\(\"([a-zA-Z0-9\\.]+)\"\\)") // resolveMessage("i18nkey")
      find(
        file,
        content,
        "throw new UserException\\(\"([a-zA-Z0-9\\.]+)\""
      ) // throw new UserException("i18nkey"
      find(
        file,
        content,
        "throw new AccessException\\(\"([a-zA-Z0-9\\.]+)\""
      ) // throw new AccessException("i18nkey"
      find(file, content, "I18N_KEY_[A-Z0-9_]+ = \"([a-zA-Z0-9\\.]+)\"") // I18N_KEY_... = "i18nKey"
      find(file, content, "new Holiday\\(\"([a-zA-Z0-9\\.]+)\"") // new Holiday("i18nKey"
      find(
        file, content,
        "MessageAction.getForwardResolution\\([a-zA-Z0-9_\"\\.]+,\\s+\"([a-zA-Z0-9\\.]+)\"\\)"
      ) // MessageAction.getForwardResolution(...,
      // "i18nkey");
      if (file.path.endsWith(PATH_DAYHOLDER) == true) {
        for (key in DayHolder.DAY_KEYS) {
          add("calendar.day.$key", file)
          add("calendar.shortday.$key", file)
        }
      } else if (file.path.endsWith(PATH_MONTHHOLDER) == true) {
        for (key in MonthHolder.MONTH_KEYS) {
          add("calendar.month.$key", file)
        }
      } else if (file.path.endsWith(PATH_MENU_ITEM_DEF) == true) {
        for (menuItem in MenuItemDefId.values()) {
          add(menuItem.i18nKey, file)
        }
      } else if (file.name.endsWith("Page.java") == true
        && (content.contains("extends AbstractListPage") == true
            || content.contains("extends AbstractEditPage") == true)
      ) { // Wicket
        // Page
        var list =
          find(content, "super\\(parameters, \"([a-zA-Z0-9\\.]+)\"\\);") // super(parameters, "i18nKey");
        for (entry in list) {
          add("$entry.title.edit", file)
          add("$entry.title.list", file)
          add("$entry.title.list.select", file)
        }
        list = find(
          content,
          "super\\(caller, selectProperty, \"([a-zA-Z0-9\\.]+)\"\\);"
        ) // super(caller, selectProperty,
        // "i18nKey");
        for (entry in list) {
          add("$entry.title.edit", file)
          add("$entry.title.list", file)
          add("$entry.title.list.select", file)
        }
      }
    }
  }

  @Throws(IOException::class)
  private fun parseKotlin(path: Path) {
    val files = listFiles(path, "kt")
    for (file in files) {
      val content = getContent(file)
      find(file, content, "translate\\(\"([a-zA-Z0-9\\.]+)\"\\)") // translate("i18nKey")
      //find(file, content, "translateMsg\\(\"([a-zA-Z0-9\\.]+,") // translateMst("i18nKey",...)
      //find(file, content, "addTranslations\\(\"([a-zA-Z0-9\\.]+)\"\\)") // addError("i18nkey")
    }
  }

  private fun find(
    file: File, content: String,
    regexp: String,
    prefix: String? = null
  ) {
    val list = find(content, regexp)
    for (entry in list) {
      val key = if (prefix != null) prefix + entry else entry
      add(key, file)
    }
  }

  private fun find(file: File, content: String, regexp: String) {
    find(content, regexp).forEach {
      add(it, file)
    }
  }

  private fun find(content: String, regexp: String): List<String> {
    val result: MutableList<String> = ArrayList()
    val p = Pattern.compile(regexp, Pattern.MULTILINE) // Compiles regular expression into Pattern.
    val m = p.matcher(content)
    while (m.find()) {
      result.add(m.group(1))
    }
    return result
  }


  private fun add(key: String, file: File) {
    add(key, file.name)
  }

  private fun add(key: String, location: String) {
    val info = ensureI18nKeyInfo(key)
    //println("$key: $location, en=${info.translation}, de=${info.translationDE}")
    info.addLocation(location)
  }

  private fun ensureI18nKeyInfo(key: String): I18nKeyInfo {
    var info = i18nKeyMap[key]
    if (info == null) {
      info = I18nKeyInfo(key)
      i18nKeyMap[key] = info
    }
    return info
  }

  private fun listFiles(path: String, suffix: String): Collection<File> {
    return FileUtils.listFiles(File(path), arrayOf(suffix), true)
  }

  private fun listFiles(path: Path, suffix: String): Collection<File> {
    return FileUtils.listFiles(path.toFile(), arrayOf(suffix), true)
  }

  @Throws(IOException::class)
  private fun getContent(file: File): String {
    return FileUtils.readFileToString(file, "UTF-8")
  }

  private val basePath: String
    get() {
      return System.getProperty("user.dir")
    }

  private fun getResourceBundle(bundleName: String, locale: Locale?): ResourceBundle {
    return if (locale != null) ResourceBundle.getBundle(bundleName, locale) else ResourceBundle.getBundle(bundleName)
  }

  companion object {
    private val PATH_DAYHOLDER = getPathForClass(DayHolder::class.java)
    private val PATH_MONTHHOLDER = getPathForClass(MonthHolder::class.java)
    private val PATH_MENU_ITEM_DEF = getPathForClass(MenuItemDef::class.java)
    private const val I18N_KEYS_FILE = "src/main/resources/" + WebConstants.FILE_I18N_KEYS

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
      CreateI18nKeys().run()
    }

    private fun getPathForClass(clazz: Class<*>): String {
      return clazz.name.replace(".", "/") + ".java"
    }
  }
}
